### PR TITLE
Add a PLCR Factory contract 

### PR DESCRIPTION
**Pull request by skmgoldin**
This PR adds a PLCRFactory.sol contract with which new PLCRVoting contracts can be spawned. It also refactors the tests to rely less on the Truffle migration context, which should make the whole repo far less brittle.

TODO: Add tests for PLCRFactory.sol

### CODE REVIEW

File: contracts/PLCRFactory.sol [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    plcr.token = new EIP20(_supply, _name, _decimals, _symbol);
+    plcr.token.transfer(msg.sender, _supply);
+
+    plcr.plcr = PLCRVoting(pf.createProxy(canonizedPLCR, ""));
+    plcr.plcr.init(plcr.token);
+
+    emit newPLCR(plcr.creator, plcr.token, plcr.plcr, plcrNonce);
+
+    plcrNonce++;
+    return plcrNonce - 1;
```
Comment (kangarang) [inline contracts/PLCRFactory.sol]: same thing here?

`return plcrNonce++;`

File: contracts/ProxyFactory.sol [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        proxyContract = createProxyImpl(_target, _data);
+
+        ProxyDeployed(proxyContract, _target);
+    }
+    
+    function createProxyImpl(address _target, bytes _data)
+        internal
+        returns (address proxyContract)
+    {
+        assembly {
```
Comment (kangarang) [inline contracts/ProxyFactory.sol]: let's learn more about what this code does!

File: test/revealVote.js [hunk 3]
```diff
@@ -96,15 +111,14 @@ contract('PLCRVoting', (accounts) => {
...
       const options = utils.defaultOptions();
       options.actor = alice;
 
-      const pollID = await utils.startPollAndCommitVote(options);
+      const pollID = await utils.startPollAndCommitVote(options, plcr);
       await utils.increaseTime(new BN(options.commitPeriod, 10).toNumber(10));
       await utils.increaseTime(new BN(options.revealPeriod).add(new BN('1', 10)).toNumber(10));
 
-      const pollEnded = await plcr.pollEnded.call(pollID);
+      const pollEnded = await plcr.pollEnded.call(pollID, plcr);
```
Comment (kangarang) [inline test/revealVote.js]: has the API for `pollEnded` changed?

File: contracts/PLCRFactory.sol [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+import "./ProxyFactory.sol";
+
+contract PLCRFactory {
+
+  event newPLCR(address creator, EIP20 token, address plcr, uint plcrNonce);
+
+  struct PLCR {
+    address creator;
+    EIP20 token;
+    PLCRVoting plcr;
```
Comment (kangarang) [inline contracts/PLCRFactory.sol]: should this be called `voting`?

File: contracts/PLCRFactory.sol [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+  }
+
+  uint plcrNonce;
+
+  ProxyFactory pf;
```
Comment (kangarang) [inline contracts/PLCRFactory.sol]: perhaps `proxyFactory`?

**Comment by Kmoneal** [comment]
Just a couple things but other than that, looks good!

File: contracts/PLCRFactory.sol [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    canonizedPLCR = new PLCRVoting();
+    proxyFactory = new ProxyFactory();
+  }
+
+  /*
+  @dev deploys and initializes a new PLCRVoting contract that consumes a token at an address
+  supplied by the user.
+  @param _token an EIP20 token to be consumed by the new PLCR contract
+  */
+  function newPLCRBYOToken(EIP20 _token) public returns (PLCRVoting) {
```
Comment (kangarang) [inline contracts/PLCRFactory.sol]: could this be simplified?

```sol
function newPLCRBYOToken(EIP20 _token) public returns (PLCRVoting) {
    address creator = msg.sender;
    EIP20 token = _token;
    PLCRVoting plcr = PLCRVoting(proxyFactory.createProxy(canonizedPLCR, ""));

    plcr.init(token);
    emit newPLCR(creator, token, plcr);
    return plcr;
}
```

or is the current implementation more efficient?

File: contracts/PLCRFactory.sol [hunk 7]
```diff
@@ -0,0 +1,61 @@
+pragma solidity ^0.4.20;
+
+import "tokens/eip20/EIP20.sol";
+import "./PLCRVoting.sol";
+import "./ProxyFactory.sol";
+
+contract PLCRFactory {
+
+  event newPLCR(address creator, EIP20 token, PLCRVoting plcr);
```
Comment (kangarang) [inline contracts/PLCRFactory.sol]: does this emit the entire contract? 

is it more efficient to emit the address?
```sol
event newPLCR(address creator, address token, address plcr);
```

**Pull request merged**